Switch single tap and long tap behaviour

strump opened the pull request:
Tap on any place - select point.
Long tap on POI - select POI.
Long tap on free space - deselect.

- Closes https://github.com/organicmaps/organicmaps/issues/1894
- Closes #1105 


https://github.com/organicmaps/organicmaps/assets/720808/164f2350-79ba-46cf-a3b9-dcdf2032049c

## Update 2024-04-29

Changed behaviour of building selection: tap on a building centroid (around 5m) then centroid is selected, otherwise tap position is selected.

https://github.com/organicmaps/organicmaps/assets/720808/87096be8-381c-42de-93d2-5d8ad1827344

## Update 2024-05-30

`Android`: Added toast for the first 2 long taps:
`Update 2024-06-06`: toast is shown only once.

https://github.com/organicmaps/organicmaps/assets/720808/7b43e046-cef0-4df3-bdf2-4aa9a0601879

pastk commented:
POIs should be selected by a short tap.

Misalf-git commented:
I guess activating "fullscreen" mode (which I often activate unintentionally) would change somehow?

biodranik commented:
@strump thanks for the change! It needs a bit of polishing to be ready for the release:

1. The single tap on a feature should behave exactly like it was behaving before, and select that feature.
2. A long tap's implementation to select any place should be called only if (1) failed and nothing was selected.
3. A long tap should hide the interface anywhere, even if it's a long tap on a POI.

biodranik commented:
@strump now a single tap on the empty space activates the full-screen mode as before...

pastk commented:
> 
> There is now a different behavior on a long tap: for pois, it selects them, for empty areas the map is switched to a full screen.

Yeap and IMO it doesn't make sense (why select POIs if its easier with a short tap?) and is not convenient.
It'd be better if a long tap just switched the UI always.

biodranik commented:
Leaving a way to select the long-tapped place is important IMO, for example, to set a custom bookmark somewhere in a larger area object. Or to make adding places easier with a pre-selection.

So the only behavior change with the master branch can be the removal of the full-screen switch from a single tap on the empty space, and selecting that empty space instead, exactly like the long-tap does in the master.

But then we need another way to enable full-screen mode somehow when it's necessary. Any ideas? 
CC @euf

pastk commented:
> But then we need another way to enable full-screen mode somehow when it's necessary. Any ideas?

A small corner button, e.g.
![image](https://github.com/organicmaps/organicmaps/assets/18434508/dc0aa045-03d9-4a4c-9248-359432b7060c)

biodranik commented:
A button visible in the full-screen mode to disable full-screen would look strange )

pastk commented:
Mapy.cz does it, it looks good (they use a round button in the corner).
I think I saw a corner button like I drew in some map app before, maybe in Guru.

strump commented:
Changed behaviour of building selection: tap on a building centroid (around 5m) then centroid is selected, otherwise tap position is selected.

But when user taps on map we search for a poi in radius 20px (actual radius in meters depends on zoom). In this PR I use fixed distance to a building centroid - 5m (doesn't depend on zoom). Is it ok?

pastk commented:
> 
> But when user taps on map we search for a poi in radius 20px (actual radius in meters depends on zoom). In this PR I use fixed distance to a building centroid - 5m (doesn't depend on zoom). Is it ok?

IMHO zoom-dependent threshold is more logical (though the current value is too high imho). If its a fixed distance threshold then on lower zooms it will be very hard to select POI/centroid, right?
But thats just my thoughts so far, let's test it in action!

strump commented:
Currently tap on map selects objects according to priorities:

1. Current position if it's withing 20px from tap position (see `FrontendRenderer::OnTap`)
2. Visible POI or user mark in 20px rect from tap position (see `FrontendRenderer::GetVisiblePOI`)
3. GPX track on long tap in 20px rect (see `Framework::FindTrackInTapPosition`)
4. Building within 1.1m rect (see `Framework::FindBuildingAtPoint`)
5. Any other object - road, fence, park, lake, etc (see `Framework::FillPointInfo`)

pastk commented:
Thanks for this analysis!

>     1. Current position if it's withing 20px from tap position (see `FrontendRenderer::OnTap`)
> 
>     2. Visible POI or user mark in 20px rect from tap position (see `FrontendRenderer::GetVisiblePOI`)

Let's try to reduce these both thresholds a little to e.g. 15px. Now the snapping is too strong.

>     3. GPX track on long tap in 20px rect (see `Framework::FindTrackInTapPosition`)

IMO it should work with a short tap. (and 15px also)

>     4. Building within 1.1m rect (see `Framework::FindBuildingAtPoint`)

You mean 1.1m from the centroid, right? It'd be great to use a pixel threshold here too.

> This function has `constexpr kSelectRectWidthInMeters = 1.1` for point search radius, and `constexpr kMetersToLinearFeature = 3` for line search radius, and non-constant argument `double toleranceInMeters` for areas search radius (0.0 by default). This looks strange. Should we have all three radiuses as function arguments?
> 
> Function `indexer::ForEachFeatureAtPoint` is used by editor, so changing it would affect edits uploading to OSM. How to improve step 5 to simplify roads selection? Introduce new function `indexer::ForEachFeatureAtPoint_V2` with customizable search radiuses for each feature type?

Maybe it'd be good to examine in more detail how its used in the editor, e.g. it shouldn't care about the lines..

Anyway the idea with passing threshold values looks good to me,
But we need pixel threshold for tapping and meters for the editor. So depending on the implementation a separate function might be the best solution.

strump commented:
@pastk  20px radius I mentioned is configured in steps 1-3 is configured in `VisualParams::GetTouchRectRadius()`. So we can adjust it to 15px there.

Also AFAIK editor uses `indexer::ForEachFeatureAtPoint()` function to merge user edits with MWM data. Each modified object (POI or building) should be matched to MWM feature.

biodranik commented:
Thanks! Looks like the priorities should be fixed too:
1. Current position
2. Bookmark or user-created/imported track
3. POI
4. Building
5. Anything else.

pastk commented:
>     2. Bookmark or user-created/imported track
> 
>     3. POI

Then it might get hard to select a POI along a user track?
The track is long, so usually there is no issue to find an easily selectable stretch.

biodranik commented:
> Then it might get hard to select a POI along a user track?
The track is long, so usually there is no issue to find an easily selectable stretch.

Good point, we can test that case specifically.

pastk commented:
A related bug.
A further service road is selected though the tram line is closer.
![image](https://github.com/organicmaps/organicmaps/assets/18434508/d0d534f3-3332-4cf4-9bed-83f40a8d6eb4)

Probably the logic returns the first match from the given tap rect/radius instead of sorting matches by distance.

pastk commented:
I've tested the current state of the PR and its quite good!

Short tapping works nicely and logical overall.

IMHO the only mandatory thing is 
- make long tap to switch the interface only, it shouldn't select anything.

After that it should be ready for another beta.

Less important things that could be done separately or later:
- tune/reduce snapping thresholds as snapping is too strong now
- make building centroid snapping use pixel threshold instead of distance (I've tested the current distance-based version and it is inferior - the snapping is super strong on high zooms and weak on lower zooms)
- the closest feature should be selected if there are several features inside the selection threshold

Changing the interface switching UX or long tap action, etc. should be brainstormed further and done separately.

pastk commented:
> * make building centroid snapping use pixel threshold instead of distance (I've tested the current distance-based version and it is inferior - the snapping is super strong on high zooms and weak on lower zooms)

Actually judging by the current behavior it seems that when a building has a caption (e.g. housenumber) or an icon (a poi tagged on the building polygon) then a pixel threshold for POIs works also. So in such cases snapping works fine even on low zooms. But on high zooms the distance based threshold makes the snapping zone too wide.

When a building doesn't have a caption/icon then I don't see any point in snapping to its centroid. Any tap inside the polygon selects the building anyway (as can be seen both in PP and in the editor). Snapping to the centroid makes sense only to visually "select" the displayed icon/caption if any.

So my suggestion is to remove the special building snapping logic completely, it should just use generic POI snapping and area selection. It works like this already for other areal POIs like universities, hospitals etc. and it works fine.

strump commented:
> * make long tap to switch the interface only, it shouldn't select anything.

I changed long tap behaviour. Not it only shows/hides UI. In navigation mode nothing happens on long tap.
Please test and tell your impression.

pastk commented:
> 
> I changed long tap behaviour. Not it only shows/hides UI. In navigation mode nothing happens on long tap. Please test and tell your impression.

Long tap works well!

But single tap stopped selecting lines and areas, it switches the UI on/off again :)

strump commented:
Fixed single-tap selection of lines.
Fixed tap circle position inside buildings: now if distance from tap point to a building centroid is less than 10 px then centroid is selected.

pastk commented:
IMO the PR should go to the next beta version.

biodranik commented:
@strump is it hard to add a toast when the interface is hidden, and display it only 2 times?
"Long-tap the map again to see the interface"

pastk commented:
In the future I think it'll be better to use long taps for something more useful (perhaps tailored to power users).

Interface show/hide could be done some other way (my suggestion is a small corner button, but maybe its too much for such a niche function).

biodranik commented:
Yes, we can show some kind of useful context menu for example. Or enable advanced features.

strump commented:
@biodranik added a toast to Android. See video in PR description

biodranik commented on data/strings/strings.txt:
```diff
@@ -7416,6 +7416,11 @@
+  [long_tap_toast]
+    comment = Toast text when user hides UI with long tap
+    tags = android
+    en = Long-tap the map again to see the interface
```
What do you think about these translations @organicmaps/translations

```suggestion
    en = Long-tap on the map again to see the interface
    af = Langtik weer op die kaart om die koppelvlak te sien
    ar = اضغط مطولاً على الخريطة مرة أخرى لرؤية الواجهة
    az = İnterfeysi görmək üçün yenidən xəritəyə uzun müddət toxunun
    be = Яшчэ раз доўга націсніце на карту, каб убачыць інтэрфейс
    bg = Докоснете отново картата, за да видите интерфейса
    ca = Toqueu llargament el mapa de nou per veure la interfície
    cs = Dlouhým klepnutím na mapu znovu zobrazíte rozhraní.
    da = Tryk længe på kortet igen for at se grænsefladen
    de = Tippe erneut lange auf die Karte, um das Interface zu sehen
    el = Πατήστε ξανά παρατεταμένα στο χάρτη για να δείτε το περιβάλλον εργασίας.
    es = Vuelve a pulsar prolongadamente sobre el mapa para ver la interfaz
    et = Puudutage pikalt kaardil uuesti, et näha kasutajaliidest.
    eu = Sakatu luze berriro mapan interfazea ikusteko
    fa = دوباره روی نقشه ضربه طولانی بزنید تا رابط کاربری را ببینید
    fi = Näytä käyttöliittymä napauttamalla karttaa uudelleen pitkään.
    fr = Appuie à nouveau longuement sur la carte pour voir l'interface.
    he = הקש ארוכות שוב על המפה כדי לראות את הממשק
    hi = इंटरफ़ेस देखने के लिए मानचित्र पर फिर से देर तक टैप करें
    hu = Érintse meg újra hosszan a térképet a kezelőfelület megjelenítéséhez.
    id = Ketuk lama pada peta sekali lagi untuk melihat antarmuka
    it = Tocca di nuovo a lungo sulla mappa per visualizzare l'interfaccia
    ja = インターフェイスを見るには、もう一度地図をロングタップする。
    ko = 지도를 다시 길게 탭하면 인터페이스가 표시됩니다.
    lt = Dar kartą ilgai bakstelėkite žemėlapį, kad pamatytumėte sąsają
    mr = इंटरफेस पाहण्यासाठी पुन्हा नकाशावर दीर्घ टॅप करा
    nb = Trykk lenge på kartet igjen for å se grensesnittet
    nl = Tik nogmaals lang op de kaart om de interface te zien
    pl = Ponownie dotknij długo mapy, aby wyświetlić interfejs
    pt = Toca longamente no mapa novamente para veres a interface
    pt-BR = Toque longamente no mapa novamente para ver a interface
    ro = Atingeți lung pe hartă din nou pentru a vedea interfața
    ru = Долгое нажатие на карту вернёт интерфейс обратно
    sk = Opätovným dlhým ťuknutím na mapu zobrazíte rozhranie
    sv = Tryck länge på kartan igen för att se gränssnittet
    sw = Gusa kwa muda mrefu kwenye ramani tena ili kuona kiolesura
    th = แตะยาวบนแผนที่อีกครั้งเพื่อดูอินเทอร์เฟซ
    tr = Arayüzü görmek için haritaya tekrar uzun dokunun
    uk = Ще раз потримайте палець на карті, щоб побачити інтерфейс
    vi = Nhấn và giữ lại vào bản đồ để xem giao diện
    zh-Hans = 再次长按地图查看界面
    zh-Hant = 再次長按地圖即可看到介面
```

NeatNit replied:
`he`:
לחץ לחיצה ארוכה במפה כדי להציג את הממשק

I used 'show' instead of 'see', feels more appropriate. Show as in the opposite of hide.

matheusgomesms replied:
Please make PT and PT-BR equal to: _Dê um toque longo no mapa novamente para ver a interface_

biodranik replied:
Thanks, updated the suggestion above. @strump can you please integrate it?

strump replied:
Added translations from suggestions.
GitHub refuses to apply through UI saying "Sorry, the diff is outdated".

biodranik commented on data/strings/strings.txt:
```diff
@@ -7416,6 +7416,11 @@
     zh-Hant = 以 8 字形移動手機以校準指南針並固定地圖上的箭頭方向
 
+  [long_tap_toast]
+    comment = Toast text when user hides UI with long tap
```
```suggestion
    comment = Toast text when user hides UI with a long tap anywhere on the map
```

biodranik commented on map/framework.cpp:
```diff
@@ -1991,8 +1992,12 @@ void Framework::OnTapEvent(place_page::BuildInfo const & buildInfo)
-
-  if (placePageInfo)
+    
+  if (buildInfo.m_isLongTap) {
```
```suggestion
  if (buildInfo.m_isLongTap)
  {
```

biodranik reviewed (commented):
@vng are you ok with this implementation? Did you test it in the latest alpha/beta?

@kirylkaveryn would it be hard to add a similar toast on iOS (show it only when the interface was hidden)?

Maybe showing toast once would be enough to educate users, to avoid overcomplicating the implementation, WDYT?

kirylkaveryn commented:
> @kirylkaveryn would it be hard to add a similar toast on iOS (show it only when the interface was hidden)?

Sure! I'll add this toast ASAP

pastk commented:
Follow-up tickets:
- https://github.com/organicmaps/organicmaps/issues/8378
- https://github.com/organicmaps/organicmaps/issues/8377

It'd be great to fix those and release together with this PR.

pastk reviewed (commented):
```
    // UI is always notified even if empty map is tapped,
    // because empty tap event switches on/off full screen map view mode.
    DeactivateMapSelection(true /* notifyUI */);
```
Outdated comment or the whole block?

biodranik commented on android/app/src/main/java/app/organicmaps/MwmActivity.java:
```diff
@@ -1264,6 +1266,16 @@ private boolean isFullscreen()
 
+  private void showFullscreenToastIfNeeded()
+  {
+    // Show the toast only 2 times so new behaviour doesn't confuse users
+    if (Config.longTapToastShowsCount(this) < 2)
+    {
+      Toast.makeText(this, R.string.long_tap_toast, Toast.LENGTH_LONG).show();
+      Config.incLongTapToastShowsCount(this);
+    }
```
After some testing, a one-time long-length toast looks good enough. Two toasts look more annoying, it feels like the toast will always be displayed on entering fullscreen. So let's do it only once (use a bool).
```suggestion
    // Show the toast only 2 times so new behaviour doesn't confuse users
    if (Config.longTapToastShowsCount(this) < 2)
    {
      Toast.makeText(this, R.string.long_tap_toast, Toast.LENGTH_LONG).show();
      Config.incLongTapToastShowsCount(this);
    }
```

@kirylkaveryn can you please add commits to also show a one-time toast on iOS (on a separate branch, so @strump can then cherry-pick your iOS commits)?

kirylkaveryn replied:
Done!

biodranik commented on drape_frontend/frontend_renderer.cpp:
```diff
@@ -1987,6 +1987,13 @@ void FrontendRenderer::OnTap(m2::PointD const & pt, bool isLongTap)
   m2::PointD mercator = screen.PtoG(pxPoint2d);
+
+  // Long tap should show/hide interface. No need to detect tapped features
+  if (isLongTap) {
```
```suggestion
  if (isLongTap)
  {
```

strump replied:
Fixed formatting

biodranik commented on drape_frontend/frontend_renderer.cpp:
```diff
@@ -1987,6 +1987,13 @@ void FrontendRenderer::OnTap(m2::PointD const & pt, bool isLongTap)
   m2::PointD const pxPoint2d = screen.P3dtoP(pt);
   m2::PointD mercator = screen.PtoG(pxPoint2d);
+
+  // Long tap should show/hide interface. No need to detect tapped features
```
```suggestion
  // Long tap should show/hide the interface. There is no need to detect tapped features.
```

biodranik replied:
@strump should this check be done at the very beginning of this method?

biodranik commented on map/framework.cpp:
```diff
@@ -2166,18 +2172,30 @@ std::optional<place_page::Info> Framework::BuildPlacePageInfo(
+      // Calculate distance between building centroid and tap point
+      double distancePx = (m_currentModelView.GtoP(buildInfo.m_mercator) - m_currentModelView.GtoP(outInfo.GetMercator())).Length();
+      // If user taps on a building far from it's centroid (>10px) move selection circle to tap position
+      if (distancePx > kBuildingCentroidThreshold * df::VisualParams::Instance().GetVisualScale())
```
@vng can you please review this change?

pastk replied:
I think this code is not needed at all, please check https://github.com/organicmaps/organicmaps/pull/7876#discussion_r1627355744

biodranik commented on map/framework.cpp:
```diff
@@ -2166,18 +2172,30 @@ std::optional<place_page::Info> Framework::BuildPlacePageInfo(
+    if (isBuildingSelected)
+    {
+      // Calculate distance between building centroid and tap point
+      double distancePx = (m_currentModelView.GtoP(buildInfo.m_mercator) - m_currentModelView.GtoP(outInfo.GetMercator())).Length();
```
nit:
```suggestion
      double const distancePx = (m_currentModelView.GtoP(buildInfo.m_mercator) - m_currentModelView.GtoP(outInfo.GetMercator())).Length();
```

pastk commented on map/framework.cpp:
```diff
@@ -2166,18 +2172,30 @@ std::optional<place_page::Info> Framework::BuildPlacePageInfo(
-    if (buildInfo.m_isLongTap)
-      outInfo.SetMercator(buildInfo.m_mercator);
+
+    if (isBuildingSelected)
+    {
+      // Calculate distance between building centroid and tap point
+      double distancePx = (m_currentModelView.GtoP(buildInfo.m_mercator) - m_currentModelView.GtoP(outInfo.GetMercator())).Length();
+      // If user taps on a building far from it's centroid (>10px) move selection circle to tap position
+      if (distancePx > kBuildingCentroidThreshold * df::VisualParams::Instance().GetVisualScale())
+        outInfo.SetMercator(buildInfo.m_mercator);
+    }
```
```suggestion
    if (isBuildingSelected)
      outInfo.SetMercator(buildInfo.m_mercator);
```

If a building has something visual at its centroid (a POI, a housenumber..) then the selection will be snapped to it already via the code before this one.
If there is nothing visual at the centroid, then why snap to it at all? Leave the selection cirlce at the tapped point.
WDYT?

(I've tested this change and it works fine for me)

(and when https://github.com/organicmaps/organicmaps/issues/4998 is fixed properly then we'll be able to remove all "building"-specific logic from here, incl. FindBuildingAtPoint())

pastk replied:
Also your current snapping code doesn't work at z14-15 anyway (see https://github.com/organicmaps/organicmaps/pull/7876#issuecomment-2117967830), so my change will make it more consistent across zoom levels, actually.

strump replied:
Accepted your solution. It works well

kirylkaveryn commented:
@strump here you can cherry-pick the toast for ios
Pplease add the `ios` tag to the strings before regen.

https://github.com/organicmaps/organicmaps/commit/5a67cab777d9483a740e624857d186fa72578f1d

strump commented:
@kirylkaveryn many thanks. Cherry-picked your commit and regenerated iOS strings

jimcarst commented:
translations ok for nl

pastk reviewed (approved):
Awesome!

biodranik commented on android/app/src/main/java/app/organicmaps/util/Config.java:
```diff
@@ -396,6 +397,16 @@ public static void setFirstStartDialogSeen(@NonNull Context context)
         .apply();
   }
 
+  public static boolean wasLongTapToastShown(@NonNull Context context) {
```
```suggestion
  public static boolean wasLongTapToastShown(@NonNull Context context)
  {
```

biodranik commented on android/app/src/main/java/app/organicmaps/util/Config.java:
```diff
@@ -396,6 +397,16 @@ public static void setFirstStartDialogSeen(@NonNull Context context)
+    return MwmApplication.prefs(context).getBoolean(KEY_PREF_LONG_TAP_TOAST_SHOWN, false);
+  }
+
+  public static void setLongTapToastShown(@NonNull Context context, Boolean newValue) {
```
```suggestion
  public static void setLongTapToastShown(@NonNull Context context, Boolean newValue)
  {
```

biodranik commented on iphone/Maps/Classes/CustomViews/MapViewControls/SideButtons/MWMSideButtons.mm:
```diff
@@ -136,5 +137,20 @@ - (void)setZoomHidden:(BOOL)zoomHidden
+  return [self.sideView setHidden:hidden animated:YES];
+}
+
+- (void)showLongTapToShowSideButtonsToastOnFirstHiding {
```
```suggestion
- (void)showLongTapToShowSideButtonsToastOnFirstHiding
{
```

biodranik reviewed (commented):
Thanks! For some reason, when I test it on Mac on Simulator, not every click is handled (randomly). @kirylkaveryn @strump did you observe a similar behavior?

biodranik commented:
@strump can you please make a clean commit history by uniting logical changes into a few commits?

strump commented:
@biodranik  rebased the branch to `master`. Re-grouped changes by drape, strings, android and ios parts.

biodranik commented on map/framework.cpp:
```diff
@@ -1992,7 +1993,12 @@ void Framework::OnTapEvent(place_page::BuildInfo const & buildInfo)
-  if (placePageInfo)
+  if (buildInfo.m_isLongTap)
+  {
+    // Show or hide UI on long tap
```
```suggestion
    // Hide Place page if it is open in any mode.
    // Show or hide the main map UI on a long tap if the Place Page is hidden (UI is not hidden in the routing/navigation mode).
```

biodranik reviewed (approved):
@strump LGTM, Дякую! A few minor nits to apply into corresponding commit(s) and we're good to go!

strump commented:
Unexpected refactoring:

When I did this change in commit [de881f6](https://github.com/organicmaps/organicmaps/commit/de881f6f8d52d44e2575cb5b0b19cd5af060f363)

```diff
--- a/map/framework.cpp
+++ b/map/framework.cpp
@@ -2187,4 +2193,4 @@ std::optional<place_page::Info> Framework::BuildPlacePageInfo(
   }
-  else if (buildInfo.m_isLongTap || buildInfo.m_source != place_page::BuildInfo::Source::User)
+  else
   {
     if (isFeatureMatchingEnabled)
```

it appeared that method `Framework::BuildPlacePageInfo` always returns some value. Case when user taps on empty space is excluded. So I changed return type from `std::optional<place_page::Info>` to `place_page::Info`. Updated all places where this method is called.

I made some smoke testing and verified basic cases: poi tap, area tap, search result tap, bookmark tap, long tap. No issues. Beta testers are welcome!

biodranik reviewed (approved):
Thanks! The code looks cleaner. Can you please combine the C++ commits?

pastk commented:
IMHO it makes sense to keep this commit separately - its a non-functional and not yet widely tested change.

biodranik commented:
> IMHO it makes sense to keep this commit separate - its a non-functional and not yet widely tested change.

The change becomes widely tested after merging into the master and releasing alpha/beta. Local tests show that it works, so it's good to go. If something can be solved right here and right now, then it should be done right here and right now.

If you see any potential issues with the change, then it's better to leave a comment there.

biodranik commented:
Ok, let's leave a refactoring as a separate commit. And a million thanks to everyone involved! Let's test it in the master branch.

pastk commented on data/strings/strings.txt:
```diff
@@ -7499,6 +7499,51 @@
+  [long_tap_toast]
+    comment = Toast text when user hides UI with a long tap anywhere on the map
+    tags = android,ios
+    en = Long-tap on the map again to see the interface
```
Actually **any** tap on the map brings the interface back.
Should we change it?

biodranik replied:
A tap on the map should select POI and show a place page. It should not show back the interface in the "browse map" mode.

CC @strump
